File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> results = new ArrayList<>();
        backtrack(results, new ArrayList<>(), nums, new boolean[nums.length]);
        return results;
    }

    private void backtrack(List<List<Integer>> results, List<Integer> cur, int[] nums, boolean[] used) {
        if (cur.size() == nums.length) {
            results.add(new ArrayList<>(cur));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;

            cur.add(nums[i]);
            used[i] = true;

            backtrack(results, cur, nums, used);

            cur.remove(cur.size() - 1);
            used[i] = false;
        }
    }
}

File: 0046-permutations/0046-permutations.kt
class Solution {
    fun permute(nums: IntArray): List<List<Int>> {
        val results = mutableListOf<List<Int>>()
        backtrack(results, mutableListOf(), nums, BooleanArray(nums.size))
        return results
    }

    private fun backtrack(
        results: MutableList<List<Int>>,
        cur: MutableList<Int>,
        nums: IntArray,
        used: BooleanArray
    ) {
        if (cur.size == nums.size) {
            results.add(cur.toList())
            return
        }

        for (i in nums.indices) {
            if (used[i]) continue

            cur.add(nums[i])
            used[i] = true

            backtrack(results, cur, nums, used)

            cur.removeAt(cur.size - 1)
            used[i] = false
        }
    }
}